### PR TITLE
EVG-17980 remove the db

### PR DESCRIPTION
**Pull request by ybrill**
[EVG-17980](https://jira.mongodb.org/browse/EVG-17980)

We're all clear to stop using the db. This PR removes all references to the db.

Now that we're post the db/storage split it seemed awkward for the `FindThisThing` methods to live on the bucket struct. Instead, this PR uses the bucket kind of like a db. It's stowed away in the env and magicked up when needed. An alternative could be to pass the bucket around explicitly.

If we prefer we could not do this and stick to just removing the models package (this is mostly what a1aa719bc64df83b2e70e0cd4ef201a6ffcba757 did).

### CODE REVIEW

**Comment by ybrill** [comment]
It also bears mentioning that this PR does away with the `TestID` string alias and changes the test ID to a string since we no longer need custom bson marshalling for test IDs.

File: model/test.go [hunk 1]
```diff
@@ -1,253 +1,241 @@
...
 }
 
-// SetBSON implements the bson.Setter interface.
-// When a testIDAlias is unmarshalled from BSON the driver will call this function to
-// unmarshal into the TestID.
-func (t *testIDAlias) SetBSON(raw bson.Raw) error {
-	var testID TestID
-	err := testID.SetBSON(raw)
+// FindTestsForBuild returns all of the test metadata for the given build ID
+// from the offline blob storage bucket.
```
Comment (julianedwards) [inline model/test.go:96]: `from then pail-backed offline storage.`

File: model/test.go [hunk 2]
```diff
@@ -1,253 +1,241 @@
...
-// unmarshal into the TestID.
-func (t *TestID) SetBSON(raw bson.Raw) error {
-	var id interface{}
-	if err := raw.Unmarshal(&id); err != nil {
-		return &bson.TypeError{
-			Kind: raw.Kind,
-			Type: reflect.TypeOf(t),
-		}
+// FindTestByID returns the test metadata for the given build ID and test ID
+// from the offline blob storage bucket.
```
Comment (julianedwards) [inline model/test.go:72]: `from the pail-backed offline storage.`

File: model/log.go [hunk 3]
```diff
@@ -1,198 +1,140 @@
...
-	var globalLogsSeq bson.M
-	if globalSeqFirst == nil {
-		globalLogsSeq = bson.M{"$gte": test.Seq}
+	// Tests should never be filtered by a time range other than AllTime
+	// since we always want to capture all the lines of either a single
+	// test or all tests.
+	return NewMergingIterator(NewBatchedLogIterator(testChunks, 4, AllTime), NewBatchedLogIterator(buildChunks, 4, tr)).Stream(ctx), nil
+}
+
+func (item *LogLineItem) Color() string {
```
Comment (julianedwards) [inline model/log.go:91]: Documentation.

File: model/log.go [hunk 4]
```diff
@@ -1,198 +1,140 @@
...
+		return found[0]
 	} else {
-		globalLogsSeq = bson.M{"$gte": *globalSeqFirst}
+		return ""
 	}
+}
 
-	if globalSeqLast != nil {
-		globalLogsSeq["$lte"] = *globalSeqLast
+func (item *LogLineItem) OlderThanThreshold(previousItem interface{}) bool {
```
Comment (julianedwards) [inline model/log.go:100]: Documentation.

File: README.md [hunk 5]
```diff
@@ -15,4 +15,3 @@ Example of running resmoke with logkeeper
     # Run this from the root directory where mongodb is cloned to:
     python buildscripts/resmoke.py --suites=core --log=buildlogger  --buildloggerUrl="http://localhost:8080"
```
[the remark this answers is not in the document]
Comment (ybrill) [reply README.md]: done

File: env/environment.go [hunk 6]
```diff
@@ -22,59 +18,23 @@ func init() {
...
-
-	globalEnv.Lock()
-	defer globalEnv.Unlock()
-	globalEnv.cleanupQueue = q
 	return nil
 }
 
-// CleanupQueue returns the cached cleanup queue from the environment.
-func CleanupQueue() amboy.Queue {
+// Bucket returns the cached pail Bucket from the environment.
```
[the remark this answers is not in the document]
Comment (ybrill) [reply env/environment.go]: done

File: testutil/testutil.go [hunk 7]
```diff
@@ -1,49 +1,42 @@
...
-
-func InitDB() error {
-	connInfo := mgo.DialInfo{
-		Addrs:   []string{"localhost"},
-		Timeout: 5 * time.Second,
-	}
-	session, err := mgo.DialWithInfo(&connInfo)
-	if err != nil {
-		return errors.Wrap(err, "can't connect to the db")
+func SetBucket(t *testing.T, initDir string) func() {
```
[the remark this answers is not in the document]
Comment (ybrill) [reply testutil/testutil.go:17]: done

File: testutil/testutil.go [hunk 8]
```diff
@@ -1,49 +1,42 @@
...
-// ClearCollections clears all documents from all the specified collections,
-// returning an error immediately if clearing any one of them fails.
-func ClearCollections(collections ...string) error {
-	if env.DBName() != testDBName {
-		return errors.New("ClearCollections should only be used against the test database")
-	}
-	db, closer := db.DB()
-	defer closer()
+	return func() {
+		if originalBucket == nil {
```
[the remark this answers is not in the document]
Comment (ybrill) [reply testutil/testutil.go]: nice
done.

File: testutil/testutil.go [hunk 9]
```diff
@@ -1,49 +1,42 @@
...
+		err = bucket.Push(context.Background(), pail.SyncOptions{
+			Local:  initDir,
+			Remote: "/",
+		})
+		require.NoError(t, err)
 	}
-
-	if err = env.SetSession(session); err != nil {
-		return errors.Wrap(err, "setting session")
+	if err := env.SetBucket(&bucket); err != nil {
```
[the remark this answers is not in the document]
Comment (ybrill) [reply testutil/testutil.go]: done

File: model/build.go [hunk 10]
```diff
@@ -1,176 +1,149 @@
...
+// and build number.
+func NewBuildID(builder string, buildNum int) (string, error) {
+	hasher := md5.New()
 
-// Insert inserts the build into the builds collection.
-func (b *Build) Insert() error {
-	db, closeSession := db.DB()
-	defer closeSession()
+	// This depends on the fact that Go's JSON implementation sorts JSON
+	// keys lexicographically for maps, which ensures consistent encoding.
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/build.go:32]: I remember that!
I think is was [EVG-14848](https://jira.mongodb.org/browse/EVG-14848)?

File: model/build.go [hunk 11]
```diff
@@ -1,176 +1,149 @@
...
+	jsonMap["buildNum"] = buildNum
+	hashstring, err := json.Marshal(jsonMap)
+
+	if err != nil {
+		return "", errors.Wrap(err, "generating json to hash for build key")
+	}
 
-	return db.C(BuildsCollection).Insert(b)
+	if _, err := hasher.Write(hashstring); err != nil {
+		return "", errors.Wrap(err, "hashing json for build key")
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/build.go]: done

File: model/build.go [hunk 12]
```diff
@@ -1,176 +1,149 @@
...
-func (b *Build) Insert() error {
-	db, closeSession := db.DB()
-	defer closeSession()
+	// This depends on the fact that Go's JSON implementation sorts JSON
+	// keys lexicographically for maps, which ensures consistent encoding.
+	jsonMap := make(map[string]interface{})
+	jsonMap["builder"] = builder
+	jsonMap["buildNum"] = buildNum
+	hashstring, err := json.Marshal(jsonMap)
+
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/build.go]: done

File: model/build.go [hunk 13]
```diff
@@ -1,176 +1,149 @@
...
+	if err != nil {
+		return "", errors.Wrap(err, "generating json to hash for build key")
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/build.go]: done

File: model/build.go [hunk 14]
```diff
@@ -1,176 +1,149 @@
...
-	err := db.C(BuildsCollection).UpdateId(id, bson.M{"$set": bson.M{"failed": true}})
-	return errors.Wrapf(err, "problem setting failed state on build %v", id)
+	return errors.Wrapf(env.Bucket().Put(ctx, b.key(), bytes.NewReader(data)), "uploading metadata for build '%s'", b.ID)
 }
 
-// IncrementSequence increments the build's sequence number by the given count.
-func (b *Build) IncrementSequence(count int) error {
-	db, closeSession := db.DB()
-	defer closeSession()
+// FindBuildByID returns the build metadata for the given ID from the offline
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/build.go]: done

File: model/log.go [hunk 15]
```diff
@@ -1,198 +1,140 @@
...
+func (ll *LogLineItem) UnmarshalJSON(data []byte) error {
+	var line []interface{}
+	if err := json.Unmarshal(data, &line); err != nil {
+		return errors.Wrap(err, "unmarshaling line into array")
 	}
 
-	return info.Removed, nil
-}
+	// timeField is generated client-side as the output of python's time.time(), which returns
+	// seconds since epoch as a floating point number
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/log.go]: done

File: model/log.go [hunk 16]
```diff
@@ -1,198 +1,140 @@
...
+	timeField, ok := line[0].(float64)
+	if !ok {
+		grip.Critical(message.Fields{
+			"message": "unable to convert time field",
+			"value":   line[0],
+		})
+		timeField = float64(time.Now().Unix())
+	}
+	// extract fractional seconds from the total time and convert to nanoseconds
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/log.go]: done

File: model/log.go [hunk 17]
```diff
@@ -1,198 +1,140 @@
...
-	return MergeLogChannels(testLogs, globalLogs), nil
+	return nil
 }
 
-// MergedTetsLogs returns a channel with the test's logs merged with the concurrent global logs.
-func MergedTestLogs(test *Test) (chan *LogLineItem, error) {
-	globalLogs, err := findGlobalLogsDuringTest(test)
+// DownloadLogLines returns log lines for a given build ID and test ID. If the
+// test ID is empty, this will return all logs lines in the build.
+func DownloadLogLines(ctx context.Context, buildID string, testID string) (chan *LogLineItem, error) {
```
[the remark this answers is not in the document]
Comment (ybrill) [reply model/log.go:89]: done

File: model/test.go (hunk 1 above)
Comment (ybrill) [reply model/test.go]: done

File: model/test.go (hunk 2 above)
Comment (ybrill) [reply model/test.go]: done

File: model/log.go (hunk 3 above)
Comment (ybrill) [reply model/log.go]: done
When I figured out what it was doing I realized the function wasn't named well so I renamed it.

File: model/log.go (hunk 4 above)
Comment (ybrill) [reply model/log.go]: done
Renamed this one too.

File: model/build.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (julianedwards) [reply model/build.go:32]: Yes! Great memory!

File: env/environment.go [hunk 18]
```diff
@@ -18,7 +18,7 @@ func init() {
 	globalEnv = &environment{}
 }
 
-// SetBucket caches a pail Bucket to be available from the environment.
+// SetBucket caches a storage Bucket to be available from the environment.
```
Comment (julianedwards) [inline env/environment.go:21]: Nit: I would do lower case b `bucket`.

File: testutil/testutil.go [hunk 19]
```diff
@@ -10,6 +10,10 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// SetBucket sets the bucket in the environment to a local pail bucket
+// backed by a temporary directory.
+// If initDir is not the empty string the contents of initDir are copied into the local bucket.
+// If an error is encounter it will fail the test.
```
Comment (julianedwards) [inline testutil/testutil.go:16]: `encountered`

File: testutil/testutil.go [hunk 20]
```diff
@@ -10,6 +10,10 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+// SetBucket sets the bucket in the environment to a local pail bucket
+// backed by a temporary directory.
+// If initDir is not the empty string the contents of initDir are copied into the local bucket.
```
Comment (julianedwards) [inline testutil/testutil.go:15]: `If initDir is not empty the contents of the directory with the given path are copied to the local bucket.`

**Review by julianedwards — APPROVED** [review]
LGTM % a few small things

**Pull request merged**